File: com-core/relink_com_core.cpp
// relink-com-core (C++ build) — small standalone registration daemon,
// per relink-com-spec.md "Mode A: relink-com-core" section.
//
// Listens on 0.0.0.0:<port> (default 8445, overridable with --port),
// receives RegisterRequest packets, updates an in-memory topic->peer
// table, and replies with RegisterAck containing the current matching
// peers for the requester's topics. Never touches the per-message data
// path -- this is bootstrap-only.

#include "relink/register.hpp"
#include <cstdio>
#include <cstring>
#include <cstdlib>
#include <vector>
#include <map>
#include <set>
#include <tuple>

#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>

using namespace relink;

struct PeerEntry {
    uint32_t ip;
    uint16_t port;
    bool operator<(const PeerEntry& o) const {
        return std::tie(ip, port) < std::tie(o.ip, o.port);
    }
};

int main(int argc, char** argv) {
    uint16_t port = kComCoreDefaultPort;
    for (int i = 1; i < argc; ++i) {
        if (std::strcmp(argv[i], "--port") == 0 && i + 1 < argc) {
            port = static_cast<uint16_t>(std::atoi(argv[++i]));
        }
    }

    int sock = ::socket(AF_INET, SOCK_DGRAM, 0);
    if (sock < 0) { std::perror("socket"); return 1; }

    struct sockaddr_in addr{};
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = INADDR_ANY;
    addr.sin_port = htons(port);
    if (::bind(sock, reinterpret_cast<struct sockaddr*>(&addr), sizeof(addr)) < 0) {
        std::perror("bind");
        return 1;
    }

    std::printf("relink-com-core (C++) listening on 0.0.0.0:%u\n", port);

    // topic_id -> set of peers registered for it
    std::map<uint16_t, std::set<PeerEntry>> table;

    uint8_t recv_buf[2048];
    uint8_t send_buf[8192];

    while (true) {
        struct sockaddr_in src{};
        socklen_t src_len = sizeof(src);
        ssize_t n = ::recvfrom(sock, recv_buf, sizeof(recv_buf), 0,
                                reinterpret_cast<struct sockaddr*>(&src), &src_len);
        if (n <= 0) continue;

        DecodedRegisterRequest req{};
        if (decode_register_request(recv_buf, static_cast<size_t>(n), &req) != RegisterDecodeResult::Ok) {
            std::fprintf(stderr, "relink-com-core: dropped malformed RegisterRequest\n");
            continue;
        }

        // Update table with this node's info for each topic it declared.
        PeerEntry self{req.node_ip, req.node_port};
        for (uint16_t i = 0; i < req.topic_count; ++i) {
            uint16_t topic = register_request_topic_at(req, i);
            table[topic].insert(self);
        }

        // Build the peer list: every (ip,port,topic) for the requester's
        // topics, excluding the requester's own entry.
        std::vector<RegisterAckPeer> peers;
        for (uint16_t i = 0; i < req.topic_count; ++i) {
            uint16_t topic = register_request_topic_at(req, i);
            for (const auto& peer : table[topic]) {
                if (peer.ip == self.ip && peer.port == self.port) continue;
                peers.push_back(RegisterAckPeer{peer.ip, peer.port, topic});
            }
        }

        size_t out_len = 0;
        auto er = encode_register_ack(0, peers.data(), static_cast<uint16_t>(peers.size()),
                                       send_buf, sizeof(send_buf), &out_len);
        if (er != RegisterEncodeResult::Ok) {
            std::fprintf(stderr, "relink-com-core: failed to encode ack (too many peers?)\n");
            continue;
        }

        ::sendto(sock, send_buf, out_len, 0,
                 reinterpret_cast<struct sockaddr*>(&src), src_len);

        struct in_addr ia{};
        ia.s_addr = htonl(req.node_ip);
        std::printf("relink-com-core: registered %s:%u (%u topics), replied with %zu peers\n",
                    inet_ntoa(ia), req.node_port, req.topic_count, peers.size());
    }

    ::close(sock);
    return 0;
}

File: com-core/relink_com_core.py
#!/usr/bin/env python3
"""relink-com-core (Python build) -- small standalone registration daemon,
per relink-com-spec.md "Mode A: relink-com-core" section.

Must produce byte-identical RegisterRequest/RegisterAck wire packets to
the C++ build (com-core/relink_com_core.cpp) -- same struct layout,
little-endian, no framework beyond the standard library (socket + struct),
per the spec's "no heavy serialization" rule applied to every language.

Wire layout (matches relink/include/relink/register.hpp exactly):
    RegisterRequestHeader = "<IHH"  node_ip(u32) node_port(u16) topic_count(u16)
                            + topic_count * "<H"
    RegisterAckHeader     = "<BH"   status(u8) peer_count(u16)
                            + peer_count * RegisterAckPeer("<IHH" ip,port,topic_id)
"""
import socket
import struct
import sys

DEFAULT_PORT = 8445

REQ_HEADER_FMT = "<IHH"     # node_ip, node_port, topic_count
REQ_HEADER_LEN = struct.calcsize(REQ_HEADER_FMT)
ACK_HEADER_FMT = "<BH"      # status, peer_count
ACK_HEADER_LEN = struct.calcsize(ACK_HEADER_FMT)
PEER_FMT = "<IHH"           # ip, port, topic_id
PEER_LEN = struct.calcsize(PEER_FMT)


def decode_register_request(buf: bytes):
    if len(buf) < REQ_HEADER_LEN:
        return None
    node_ip, node_port, topic_count = struct.unpack_from(REQ_HEADER_FMT, buf, 0)
    expected = REQ_HEADER_LEN + topic_count * 2
    if len(buf) != expected:
        return None
    topics = list(struct.unpack_from("<%dH" % topic_count, buf, REQ_HEADER_LEN)) if topic_count else []
    return node_ip, node_port, topics


def encode_register_ack(status: int, peers) -> bytes:
    out = bytearray(struct.pack(ACK_HEADER_FMT, status, len(peers)))
    for ip, port, topic_id in peers:
        out += struct.pack(PEER_FMT, ip, port, topic_id)
    return bytes(out)


def main():
    port = DEFAULT_PORT
    if "--port" in sys.argv:
        port = int(sys.argv[sys.argv.index("--port") + 1])

    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sock.bind(("0.0.0.0", port))
    print(f"relink-com-core (Python) listening on 0.0.0.0:{port}", flush=True)

    table = {}  # topic_id -> set of (ip, port)

    while True:
        data, addr = sock.recvfrom(2048)
        decoded = decode_register_request(data)
        if decoded is None:
            print("relink-com-core: dropped malformed RegisterRequest", file=sys.stderr, flush=True)
            continue
        node_ip, node_port, topics = decoded
        self_entry = (node_ip, node_port)

        for topic in topics:
            table.setdefault(topic, set()).add(self_entry)

        peers = []
        for topic in topics:
            for (ip, p) in table.get(topic, ()):
                if (ip, p) == self_entry:
                    continue
                peers.append((ip, p, topic))

        ack = encode_register_ack(0, peers)
        sock.sendto(ack, addr)

        # node_ip is stored host-byte-order (see relink/register.hpp /
        # ipv4_to_host_order) -- pack big-endian to get network order for
        # inet_ntoa, regardless of this machine's own endianness.
        ip_str = socket.inet_ntoa(struct.pack(">I", node_ip))
        print(f"relink-com-core: registered {ip_str}:{node_port} "
              f"({len(topics)} topics), replied with {len(peers)} peers", flush=True)


if __name__ == "__main__":
    main()
